Clarify that `PartialOrd` does not give a partial order

Muon opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/140779)*
<!-- homu-ignore:end -->

Fixes rust-lang/rust#140654.

Mark-Simulacrum commented:
I'm going to move this to the libs-api agenda -- maybe we can find someone better suited to review this. It's not clear to me from the description in the issue (https://github.com/rust-lang/rust/issues/140654) whether this is actually a semantic change or not in what we're promising -- it seems like it is maybe not, based on the diff, but I'm a little confused why we didn't land on this when initially landing the docs here.

Muon commented:
I do not believe this a semantic change. That would require the changed sentences to have been normative, but unless `f32` and `f64` are not meant to be `PartialOrd`, that's not possible, since the `<=` relation on `f32` and `f64` is already not a partial order.

RalfJung commented:
I think this PR is a bad idea. The prefix "strict" for partial orders is used to distinguish `<` and `≤` on e.g. numbers. `PartialOrd` has both `<` and `≤`, so it doesn't make sense to say that it is specifically the strict or reflexive version of any partial order.

hanna-kruppe commented:
While the trait does not require all implementations to have reflexivity, if you do want to implement a partial order that is reflexive (e.g., subset relation) then `PartialOrd` is the right trait for it. So at least the change to the module level docs seems misleading in context:

> `Ord` and `PartialOrd` are traits that allow you to define total and **strict** partial orderings between values, respectively.

Code that is generic over `T: PartialOrd` can't assume reflexivity, and since equality is partial as well, the usual way of flipping between `<` and `<=` doesn't really work. Making this "what can I know about an arbitrary `impl PartialOrd`" angle clearer could be useful, but may be better achieved by applying the following suggestion from #140654 directly:

> The docs should make clear upfront that reflexivity is not required.

I know what reflexivity is, but I'm not confident I would infer that the passing mention of "*strict* partial order" if I didn't know it already.

(Spelling out more corollaries is useful IMO, but I haven't checked the one added here.)

Muon commented:
> I think this PR is a bad idea. The prefix "strict" for partial orders is used to distinguish `<` and `≤` on e.g. numbers.

"Partial order" on its own always refers to a relation like ≤ on numbers (i.e. reflexive, antisymmetric, and transitive), so the original sentence is definitely misleading. 

> `PartialOrd` has both `<` and `≤`, so it doesn't make sense to say that it is specifically the strict or non-strict version of any partial order.

This is an argument against having the current sentence too.

> While the trait does not require all implementations to have reflexivity, if you do want to implement a partial order that is reflexive (e.g., subset relation) then `PartialOrd` is the right trait for it. So at least the change to the module level docs seems misleading in context:

Agreed, but then I'm not sure how to concisely describe what precisely `PartialOrd` is for in that context, whereas a correct implementation of `Ord` definitely produces a total order, in that (modulo equivalence) `<=` is reflexive, antisymmetric, transitive, and connected.

> In fact you cannot say a set (here, the set of all possible values of a type) form a partial order. However, you can say a set and a binary relation form a partially ordered set.

Technically true, but I rarely see the distinction made either in writing or in speech when there is only one set and one obvious order we are discussing.

> `<=` defined as such isn't even an order (mathematically speaking).

Agreed, which is why I reported the issue and made the PR.

RalfJung commented:
> In my experience, when we say "X forms a partial order", this is short for "the usual order on X is a partial order".

I rarely encounter such terminology and would consider it sloppy. The typical thing people say in this case is "X is a poset".

I am not particularly attached to the current wording. Picking the `<=` relation as the "canonical" one and identifying `PartialOrd` with `<=` is insofar justified as it is typical convention in mathematics to assume a "<=" relation and then freely use the other symbols as well as derived notions. But ofc sans reflexivity, `PartialOrd` is not actually a partial order. I don't think there is a standard term for a relation that's transitive but may not be reflexive. The first sentence explaining `PartialOrd` should probably avoid referencing any particular mathematical notion.

BartMassey commented:
@traviscross The floating NaN example you linked is, as far as I can remember, the only irreflexive "`PartialOrd`" defined anywhere in `std`? Maybe we should instead indicate that `PartialOrd` implementations *should* be reflexive, but we carved out an exception for floating types for ergonomic reasons? I'm not sure there would be any real consequences of going this route, except better expressing intent.

Muon commented:
> I rarely encounter such terminology and would consider it sloppy. The typical thing people say in this case is "X is a poset".

Oh, it's definitely sloppy! But not any more than saying "X is a poset" :P

> Picking the `<=` relation as the "canonical" one and identifying `PartialOrd` with `<=` is insofar justified as it is typical convention in mathematics to assume a "<=" relation and then freely use the other symbols as well as derived notions. But ofc sans reflexivity, `PartialOrd` is not actually a partial order. I don't think there is a standard term for a relation that's transitive but may not be reflexive.

Well, I tried to keep the mathematical correspondence by identifying `PartialOrd` with `<` (which it does promise is a strict partial order), but it seems like it might be best to dispense with it.

> The first sentence explaining `PartialOrd` should probably avoid referencing any particular mathematical notion.

I suspect people look to `PartialOrd` mainly as a way of providing `<`, `<=`, `>` and `>=`. Perhaps the docs should begin with that in mind?

---

> Maybe we should instead indicate that `PartialOrd` implementations _should_ be reflexive, but we carved out an exception for floating types for ergonomic reasons?

I think calling out the exception would be good. However, I am worried that it might be misunderstood as a logical requirement with one weird exception. Maybe we should say something more like this:

"Despite its name, `PartialOrd` does not guarantee that `<=` is a [partial order](https://en.wikipedia.org/wiki/Partial_order), because `<=` is not required to be reflexive. But if `Eq` is also implemented, this does imply that `<=` is reflexive and thus a partial order. In other words, to require `T` to be partially ordered, write `T: PartialOrd + Eq`."

For completeness, here's the justification:

**Proposition:** In a compliant implementation of `PartialOrd`, `<=` is reflexive if and only if `==` is reflexive.
_Proof:_ Let `x` be a value of the type. If `<=` is reflexive, then `x <= x`, and hence by antisymmetry we have `x == x`, and so `==` is reflexive. If `==` is reflexive, then `x == x`, and since `x <= x` if and only if either `x < x` or `x == x`, we have `x <= x`, so `<=` is reflexive as well. ∎

RalfJung commented:
> Maybe we should instead indicate that PartialOrd implementations should be reflexive, but we carved out an exception for floating types for ergonomic reasons?

I don't think that reflects the intent of this trait. Is is indeed truly okay for for this relation to not be reflexive.

> I suspect people look to PartialOrd mainly as a way of providing <, <=, > and >=. Perhaps the docs should begin with that in mind?

Yeah that makes sense.

alex-semenyuk commented:
@Muon
Thanks for your contribution.
Form wg-triage. Any updates on this PR?

Muon commented:
I started on an updated version addressing the comments, but haven't had
the time to finish it. Hopefully I will get the chance during the public
holiday next month.

On Sun, 10 Aug 2025 at 16:59, alexey semenyuk ***@***.***>
wrote:

> *alex-semenyuk* left a comment (rust-lang/rust#140779)
> <https://github.com/rust-lang/rust/pull/140779#issuecomment-3172420603>
>
> @Muon <https://github.com/Muon>
> Thanks for your contribution.
> Form wg-triage. Any updates on this PR?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/140779#issuecomment-3172420603>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADCKOJONL2AUMQEVDCTZ3D3M3UV5AVCNFSM6AAAAAB4VN2OV2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCNZSGQZDANRQGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Muon committed:
Correct and expand docs for PartialOrd

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18074874117/job/51429898318) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/51429898318) [(plain)](https://github.com/rust-lang/rust/commit/5934edc77c9a3821d129928829a8f36f7e29f21b/checks/51429898318/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test library/core/src/unit.rs - unit::() (line 6) ... ok

failures:

---- library/core/src/cmp.rs - cmp::PartialOrd (line 1338) stdout ----
error: leading `+` is not supported
 --> library/core/src/cmp.rs:1340:9
  |
5 | let b = +0.0;
  |         ^ unexpected `+`
  |
---

error[E0689]: can't call method `to_bits` on ambiguous numeric type `{float}`
 --> library/core/src/cmp.rs:1341:14
  |
4 | let a = -0.0;
  |     - you must specify a type for this binding, like `f32`
5 | let b = +0.0;
6 | assert_ne!(a.to_bits(), b.to_bits()); // `a` and `b` are distinct because their sign bits differ.
  |              ^^^^^^^

error[E0689]: can't call method `to_bits` on ambiguous numeric type `{float}`
 --> library/core/src/cmp.rs:1341:27
  |
5 | let b = +0.0;
  |     - you must specify a type for this binding, like `f32`
6 | assert_ne!(a.to_bits(), b.to_bits()); // `a` and `b` are distinct because their sign bits differ.
  |                           ^^^^^^^

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0689`.
```

</details>
